plat/drivers/virtio/virtio_ring.c

```diff
@@ -411,12 +417,31 @@ struct virtqueue *virtqueue_create(__u16 queue_id, __u16 nr_descs, __u16 align,
 	vrq->vring_mem = NULL;
 
 	ring_size = vring_size(nr_descs, align);
+#ifdef CONFIG_LIBUKVMEM
+	struct uk_pagetable *pt = ukplat_pt_get_active();
+	__paddr_t paddr = __PADDR_ANY;
+	__vaddr_t vaddr = __VADDR_ANY;
+
+	ring_size = PAGE_ALIGN_UP(ring_size);
+
+	rc = pt->fa->falloc(pt->fa, &paddr, ring_size >> PAGE_SHIFT, 0);
+	if (unlikely(rc))
+		goto err_freevq;
+
+	rc = uk_vma_map_dma(uk_vas_get_active(), &vaddr, ring_size,
+			    PAGE_ATTR_PROT_RW, UK_VMA_MAP_POPULATE,
+			    "virtqueue", paddr);
+	if (unlikely(rc))
+		goto err_freevq;
+
+	vrq->vring_mem = (void *)vaddr;
+#else /* CONFIG_LIBUKVMEM */
 	if (uk_posix_memalign(a, &vrq->vring_mem,
```

Comment:
I think this is already not safe for `HAVE_PAGING && !LIBUKVMEM`. See the main review comment.

Reply:
Yes, this is correct. This has to be addressed in a future PR as this comes with a set of difficulties:
- We cannot simply allocate memory from the frame allocator because physical memory is allocated (almost) entirely by the heap if we have paging but no ukvmem.
- Without ukvmem we have to select a virtual address somehow for the DMA mapping. However, this can easily lead to conflicts and the solution has to cope with multiple mappings.

Since the chance is very high for the queue being allocated from a contiguous 2MB huge heap page with paging enabled, and since we already have this problem in the current version, I will not address this in this PR. I will open an issue to track this.